kmk/modules/encoder.py

```diff
@@ -0,0 +1,221 @@
+    OFF = False
+    ON = True
+
+
+class Encoder:
+    def __init__(
+        self,
+        pad_a,
+        pad_b,
+        button_pin=None,
+    ):
+        self.pad_a = self.PreparePin(pad_a)  # board pin for enc pin a
+        self.pad_a_state = False
+        self.pad_b = self.PreparePin(pad_b)  # board pin for enc pin b
+        self.pad_b_state = False
+        self.button_pin = self.PreparePin(button_pin)  # board pin for enc btn
+        self.button_state = None  # state of pushbutton on encoder if enabled
+        self.encoder_value = 0  # clarify what this value is
+        self.encoder_state = (
+            self.pad_a_state,
+            self.pad_b_state,
+        )  # quaderature encoder state
+        self.encoder_direction = None  # arbitrary, tells direction of knob
+        self.last_encoder_state = None  # not used yet
+        self.resolution = 2  # number of keys sent per position change
+        self.revolution_count = 20  # position changes per revolution
+        self.has_button = False  # enable/disable button functionality
+        self.encoder_data = None  # 6tuple containing all encoder data
+        self.position_change = None  # revolution count, inc/dec as knob turns
+        self.last_encoder_value = 0  # not used
+        self.is_inverted = False  # switch to invert knob direction
+        self.vel_mode = False  # enable the velocity output
+        self.vel_ts = None  # velocity timestamp
+        self.last_vel_ts = 0  # last velocity timestamp
+        self.encoder_speed = None  # ms per position change(4 states)
+        self.encoder_map = None
+        self.eps = EncoderPadState()
+        self.encoder_pad_lookup = {
+            False: EncoderPadState.OFF,
+            True: EncoderPadState.ON,
+        }
+
+    def __repr__(self, idx):
+        return 'ENCODER_{}({})'.format(idx, self._to_dict())
+
+    def _to_dict(self):
+        return {
+            'Encoder_State': self.encoder_state,
+            'Direction': self.encoder_direction,
+            'Value': self.encoder_value,
+            'Position_Change': self.position_change,
+            'Speed': self.encoder_speed,
+            'Button_State': self.button_state,
+        }
+
+    # adapted for CircuitPython from raspi
+    def PreparePin(self, num):
+        if num is not None:
+            pad = digitalio.DigitalInOut(num)
+            pad.direction = digitalio.Direction.INPUT
+            pad.pull = digitalio.Pull.UP
+            return pad
+        else:
+            return None
+
+    # checks encoder pins, reports encoder data
+    def report(self):
+        new_encoder_state = (
+            self.encoder_pad_lookup[int(self.pad_a.value)],
+            self.encoder_pad_lookup[int(self.pad_b.value)],
+        )
+
+        if self.encoder_state == (self.eps.ON, self.eps.ON):  # Resting position
+            if new_encoder_state == (self.eps.ON, self.eps.OFF):  # Turned right 1
+                self.encoder_direction = 'R'
```

Comment:
Could we use a bool for this, and document which is which? That should take less resources, even if only by a bit.

Reply:
Could also do something like this with a `const`. That will be easier to read, and low memory. `from micropython import const`